### DISCUSSION TITLE
How mesh density is partitioned in GeneratedMesh？

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#21426] [Q&A General] [answered]
Dear MOOSE experts,

Recently I created an initial microstructure of gradient nano-grains for grain growth and used `GeneratedMesh` for uniform meshing. Under the condition that the grain mesh of the surface layer of Level 1 is guaranteed to be reasonable (see figure.(c)), it also causes the grain grid of the coarse-grained layer of level 3 to be excessively dense (see figure.(d)), which makes dofs extremely high. 
![image](https://user-images.githubusercontent.com/56183775/175855237-e899d5e7-4b4a-4b20-8ca2-fb6b08a01ffa.png)
Figure: (a) Initial microstructure of gradient nano-grains (GNGs) by `PolycrystalVoronoi` object, (b) Global meshing graph, (c) Enlarged view of meshing at Level 1 fine-grained layer, (d) Enlarged view of meshing at Level 3 coarse-grained layer.

So what I'm asking now is if there is a way to use `GeneratedMesh` or similar meshing tool to use different mesh densities at different levels or regions.

Any suggestions or recommendations to fix the problem would be greatly appreciated.

Thank you
Wei

**Comment by GiudGiud** [comment, marked as answer]
Hello

How are the different regions defined? Cartesian regions? Or have you defined subdomains?

Guillaume

**Reply by PengWei97** [reply]
Should be a Cartesian coordinate area. Specifically, I want to divide three areas or more according to different y-axis coordinates, by the way I modified my figure.

**Reply by PengWei97** [reply]
and this is the mesh setup I'm using now,
```bash
my_nx = 25 # 30 94
my_xmax = 886 # 1200 3760 400e-1 nm
my_ny = 25 # 30 94 
my_ymax = 886 # 1200 3760 400e-1 nm
[Mesh]
  # Mesh block.  Meshes can be read in or automatically generated
  type = GeneratedMesh
  dim = 2 # Problem dimension
  nx = ${my_nx} # Number of elements in the x-direction
  ny = ${my_ny} # Number of elements in the y-direction
  xmin = 0    # minimum x-coordinate of the mesh
  xmax = ${my_xmax} # 1000 maximum x-coordinate of the mesh 2000-400 400 1600
  ymin = 0    # minimum y-coordinate of the mesh
  ymax = ${my_ymax} # 1000 maximum y-coordinate of the mesh
  elem_type = QUAD4  # Type of elements used in the mesh
  uniform_refine = 0 # Initial uniform refinement of the mesh

  parallel_type = distributed # Periodic BCs distributed replicated
[]

```

**Reply by GiudGiud** [reply]
This mesh generator will let you set different mesh fineness in different cartesian regions
https://mooseframework.inl.gov/source/meshgenerators/CartesianMeshGenerator.html

I ll just mention that if you assign blocks to regions you can also perform selective block refinements (sideset is also an option, with another MG)
using this
https://mooseframework.inl.gov/source/meshgenerators/RefineBlockGenerator.html

**Reply by PengWei97** [reply]
Ok, thanks for your answer, I will try it.

**Reply by PengWei97** [reply]
hi, @GiudGiud. According to your suggestion, I used `CartesianMeshGenerator` and `SideSetsBetweenSubdomainsGenerator` to refine the mesh, the code and mesh figure are as follows,
![image](https://user-images.githubusercontent.com/56183775/176111871-7960a2fb-6ad3-4732-8580-4088832414ea.png)

However, passing `RefineBlockGenerator` in this way does not meet my needs very well. Specifically, I want to be able to set `nx` and `ny` belonging to each block like in `GeneratedMesh`, is this possible?

wei

**Reply by GiudGiud** [reply]
Hello

The refineblockGenerator only does uniform refinement currently. 
If you want custom nx and ny you may set them directly for each block in the Cartesian mesh generator.
I recommend using parsed expressions if you want to functionalize this mesh as a function of a few refinement parameters

Guillaume

**Reply by PengWei97** [reply]
Yeah, I solved this problem by modifying `max_h_level`, the effect is as follows.
![image](https://user-images.githubusercontent.com/56183775/176807241-30f050bc-9c8f-49ea-ae22-6de88d365451.png)

Thank you so much.
wei
